internal/controllers/core/tiltfile/reconciler.go

```diff
@@ -184,6 +203,54 @@ func (r *Reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
 	return ctrl.Result{}, nil
 }
 
+type tiltfileContentsResult struct {
+	sha256   string
+	contents []byte
+}
+
+func (r *Reconciler) reconcileTiltfileContents(ctx context.Context, tf *v1alpha1.Tiltfile, tlr *tiltfile.TiltfileLoadResult) (tiltfileContentsResult, error) {
+	result := tiltfileContentsResult{sha256: tf.Status.ContentsSHA256}
+	bytes, err := os.ReadFile(tf.Spec.Path)
+	if err != nil {
+		// If the file doesn't exist, don't do any reconciliation of contents
+		// and avoid emitting a reconciler error to the UI
+		if os.IsNotExist(err) {
+			return result, nil
+		}
+
+		return result, err
+	}
+
+	fileSha256 := fmt.Sprintf("%x", sha256.Sum256(bytes))
+	contentsSha256 := fmt.Sprintf("%x", sha256.Sum256([]byte(tf.Spec.Contents)))
+	result.sha256 = fileSha256
+
+	// File different from contents; file wins by default
+	if fileSha256 != contentsSha256 {
+		result.contents = bytes
+	}
+
+	// Short-circuit here when feature is disabled to prevent modifying disk file
+	if (tlr != nil && !tlr.FeatureFlags[feature.TiltfileEditAPI]) ||
+		(tlr == nil && !r.defaults[feature.TiltfileEditAPI].Enabled) {
+		return result, nil
+	}
+
+	// File contents changed since last hash
+	if fileSha256 != tf.Status.ContentsSHA256 {
+		return result, nil
+	}
+
+	// API contents changed since last hash, update file on disk
+	if contentsSha256 != tf.Status.ContentsSHA256 {
+		result.sha256 = contentsSha256
+		result.contents = []byte{}
+		err = os.WriteFile(tf.Spec.Path, []byte(tf.Spec.Contents), 0644)
+	}
```

Comment:
I'm confused by this - `Reconcile()` uses this return result and updates `tf.Spec` if `contents` is set and `tf.Status` if `sha256` is set.

So you're writing a new file to disk, which will trigger a FW event and cause re-reconcile: that makes sense. But you're going to preemptively set the `tf.Status.ContentsSHA256` right now, even though we haven't actually used that yet.